msm/src/serialization/witness.rs

```diff
@@ -274,17 +278,49 @@ pub fn deserialize_field_element<Fp: Field, Env: InterpreterEnv<Fp>>(
 
     env.add_constraint(constraint);
 
+    // -- Start third constraint
     // FIXME: range check
-    let c12 = Env::get_column_for_msm_limb(12);
-    let c13 = Env::get_column_for_msm_limb(13);
-    let c14 = Env::get_column_for_msm_limb(14);
-    let c15 = Env::get_column_for_msm_limb(15);
-    let c16 = Env::get_column_for_msm_limb(16);
-    env.bitmask_be(&input_limb2, 19, 4, c12);
-    env.bitmask_be(&input_limb2, 34, 19, c13);
-    env.bitmask_be(&input_limb2, 49, 34, c14);
-    env.bitmask_be(&input_limb2, 64, 49, c15);
-    env.bitmask_be(&input_limb2, 79, 64, c16);
+    let c12_var = {
+        let c12 = Env::get_column_for_msm_limb(12);
+        env.bitmask_be(&input_limb2, 19, 4, c12)
+    };
+    let c13_var = {
+        let c13 = Env::get_column_for_msm_limb(13);
+        env.bitmask_be(&input_limb2, 34, 19, c13)
+    };
+
+    let c14_var = {
+        let c14 = Env::get_column_for_msm_limb(14);
+        env.bitmask_be(&input_limb2, 49, 34, c14)
+    };
+
+    let c15_var = {
+        let c15 = Env::get_column_for_msm_limb(15);
+        env.bitmask_be(&input_limb2, 64, 49, c15)
+    };
+    let c16_var = {
+        let c16 = Env::get_column_for_msm_limb(16);
+        env.bitmask_be(&input_limb2, 79, 64, c16)
+    };
+
+    // Unfolding for readability.
+    let shl_15 = Fp::from(1u128 << 15u128);
+    let mut cst = Env::constant(shl_15);
+    let mut constraint = c12_var;
+
+    constraint = constraint + c13_var * cst.clone();
+    cst = cst * Env::constant(shl_15);
+    constraint = constraint + c14_var * cst.clone();
+    cst = cst * Env::constant(shl_15);
+    constraint = constraint + c15_var * cst.clone();
+    cst = cst * Env::constant(shl_15);
+    constraint = constraint + c16_var * cst.clone();
+
+    constraint = (1..20).fold(constraint, |constraint, i| {
```

Comment:
you probably want to comment the equation in the https://github.com/o1-labs/proof-systems/tree/master/msm/src/serialization file

Reply:
It is, see second equation:
```

2. For the remaining 75 bits:

$$c_{12} + c_{13} * 2^{15} + c_{14} 2^{15 * 2} + c_{15} 2^{15 * 3} + c_{16} 2^{15 * 4} = \sum_{j = 1}^{19} b_{2, j} * 2^{4 (j - 1)}$$
```

There is a wrong index though. It is not 19 but 20. Fixing it in a following PR

Reply:
Ah no, it is correct. Starting at 0...